crates/containerd-shim-wasm/src/container/context.rs

```diff
@@ -0,0 +1,236 @@
+use std::iter::{FilterMap, Map};
+use std::path::{Path, PathBuf};
+use std::slice::Iter;
+use std::str::Split;
+
+use oci_spec::runtime::Spec;
+
+// Ideally this would be `impl Iterator<Item = (&str, &str)>`
+// but we can't return `impl ...` in traits.
+pub type EnvIterator<'a> = Map<Iter<'a, String>, fn(&String) -> (&str, &str)>;
+
+// Ideally this would be `impl Iterator<Item = PathBuf>`
+// but we can't return `impl ...` in traits.
+pub type FindInPathIterator<'a> =
+    FilterMap<Split<'a, char>, Box<dyn FnMut(&str) -> Option<PathBuf>>>;
+
+pub trait RuntimeContext {
+    // ctx.args() returns arguments from the runtime spec process field, including the
+    // path to the entrypoint executable.
+    fn args(&self) -> &[String];
+
+    // ctx.entrypoint() returns the entrypoint path from arguments on the runtime
+    // spec process field.
+    fn entrypoint(&self) -> Option<&Path>;
+
+    // ctx.module() returns the module path and exported function name to be called
+    // as a (&Path, &str) tuple, obtained from the arguments on the runtime spec process
+    // field. The first argument will be the module name and the default function name
+    // is "_start".
+    //
+    // If there is a '#' in the argument it will split the string returning the first part
+    // as the module name and the second part as the function name.
+    //
+    // example: "/app/module.wasm#function" will return the tuple
+    // (Path::new("/app/module.wasm"), "function")
+    //
+    // If there are no arguments then it will return (Path::new(""), "_start")
+    fn module(&self) -> (&Path, &str);
```

Comment:
Instead of `(&Path, &str)`, could you please create a struct that captures these two fields? Something like

```
struct WasiModule { path: Path, function: String }
```

Reply:
Done.
```rust
pub struct WasiEntrypoint {
    pub path: PathBuf,
    pub func: String,
}
```